ocrd_eval.md

```diff
@@ -0,0 +1,372 @@
+
+Example:
+
+![A Fraktur sample reading "Die Finſterniſſe des 1801ſten Jahrs"](https://pad.gwdg.de/uploads/b894049b-8d98-4fe7-ac31-71b2c9393a6c.jpg)
+
+This heading reads `Die Finſterniſſe des 1801ſten Jahrs`. The output contains `180iſten`, replacing `1` with `i`.
+
+* **insertion**: a new character is introduced in the output.
+
+Example:
+
+![A Fraktur sample reading "diese Strahlen, und"](https://pad.gwdg.de/uploads/e6b6432e-d79c-4568-9aef-15a026c05b39.jpg)
+
+This reads `diese Strahlen, und`. The output contains `Strahlen ,`, inserting a white space before the comma.
+
+CER can be calculated in several ways, depending on whether a normalized CER is used or not.
+
+Given $i$ as the number of insertions, $d$ the number of deletions, $s$ the number of substitutions and $n$ the total number of characters in a text, the CER can be obtained by
+
+$CER = \frac{i + s+ d}{n}$
+
+If the CER value is calculated this way, it represents the percentage of characters incorrectly recognized by the OCR engine. Also, we can easily reach error rates beyond 100% when the output contains a lot of insertions.
+
+The *normalized* CER tries to mitigate this effect by considering the number of correct characters, $c$:
+
+$CER_n = \frac{i + s+ d}{i + s + d + c}$
+
+In OCR-D's benchmarking we calculate the *non-normalized* CER where values over 1 should be read as 100%.
+
+###### CER Granularity
+
+In OCR-D we distinguish between the CER per **page** and the **overall** CER of a text. The reasoning behind this is that the material OCR-D mainly aims at (historical prints) is very heterogeneous: Some pages might have an almost simplistic layout while others can be highly complex and difficult to process. Providing only an overall CER would cloud these differences between pages.
+
+At this point we only provide a CER per page; an overall CER might be calculated as a weighted aggregate at a later stage.
+
+##### Word Error Rate (WER)
+
+The word error rate (WER) is closely connected to the CER. While the CER focusses on differences between characters, the WER represents the percentage of words incorrectly recognized in a text.
+
+CER and WER share categories of errors, and the WER is similarly calculated:
+
+$WER = \frac{i_w + s_w + d_w}{n_w}$
+
+where $i_w$ is the number of inserted, $s_w$ the number of substituted, $d_w$ the number of deleted and $n_w$ the total number of words.
+
+More specific cases of WER consider only the "significant" words, omitting e.g. stopwords from the calculation.
+
+###### WER Granularity
+
+In OCR-D we distinguish between the WER per **page** and the **overall** WER of a text. The reasoning here follows the one of CER granularity.
+
+At this point we only provide a WER per page; an overall WER might be calculated at a later stage.
+
+#### Bag of Words
+
+In the "Bag of Words" model a text is represented as a set of its word irregardless of word order or grammar; Only the words themselves and their number of occurence are considered.
```

Comment:
I haven't found a definition, either. Ping @cneud @kba, maybe they have stumbled over something.

Reply:
Pletschacher et al 2015 suggest they want to control for false positives:
> For the index-based **success rate** it is only important for the OCR engine to find each word at least once and not to introduce false words. The count-based **success measure** is stricter and demands the correct count of recognised words (e.g. have all occurrences of the name `Shakespeare` been found or only seven out of nine).

So whatever `success rate` is, it's not just **recall**. Could it be **precision** then? But what about false negatives – they could deviate from false positives (e.g. if the OCR tends to generate too few or too many words). For **accuracy**, you could count all words in OCR and GT minus positive deltas minus negative deltas.

Unfortunately, we don't get an implementation from PRImA (they only partly publish [FCA code](https://github.com/PRImA-Research-Lab/prima-text)).

I have found an [implementation](https://github.com/impactcentre/ocrevalUAtion/blob/84f15b894eccd05365a9116f03b5b8b97c1b74b6/src/main/java/eu/digitisation/output/ErrorMeasure.java#L135-L148) in [ocrevalUAtion](https://github.com/impactcentre/ocrevalUAtion/blob/84f15b894eccd05365a9116f03b5b8b97c1b74b6/src/main/java/eu/digitisation/document/TermFrequencyVector.java#L47-L86) though. This is not normalized (the denominator is the number of GT tokens). But even more strangely, it sums up positive and negative deltas separately, and then only picks the larger one. How can they sum up the frequencies/probabilities of false negatives and of false positives separately, pick one over the other and expect a useful result?

Another definition can be found in [digital-eval](https://github.com/ulb-sachsen-anhalt/digital-eval/blob/edc8b979b3f39551d1d6c624c5ebc04076b49a74/src/digital_eval/metrics.py#L442-L451): sum of the positive deltas divided by number of GT tokens. Fair enough, that sounds like precision.

Does anyone know other implementations?

Reply:
For ocrevalUAtion, cf. [this issue](https://github.com/impactcentre/ocrevalUAtion/issues/30).

Reply:
Ping @kba @cneud

Reply:
@bertsky What do you think of @b2m's implementation of `bag_accuracy` in https://github.com/qurator-spk/dinglehopper/pull/60/files#diff-d007b01bb64f22ad1652d54d73e80442f824f03c4460ad69fb448ce2c7a54c53R57-R96 ?

>  For accuracy, you could count all words in OCR and GT minus positive deltas minus negative deltas.

I think that's what the function does.

I'm also askin since we're using dinglehopper for the evaluations, it would be worthwile to resurrect this very good PR and use as base for BoW calculation.

Reply:
@kba that PR is a bit strange. `bag_accuracy` itself looks good, but it's used by `bag_of_words_accuracy` and `bag_of_chars_accuracy` with `Weights(1, 0, 1)` – i.e. only _deletes_, but no _inserts_ (note that _replacements_ will be zero). It's wrong to call this _accuracy_ (it lends itself to _recall_).

Also, `MetricResult.error_rate` seems to be wrong, because it uses the length of the reference as denominator (instead of length of `reference` plus length of `compared`). So `accuracy` would also be wrong.

Together these two mistakes compensate each other, so that implementation calculates a valid **recall** (complement of _false negative rate_).